include/envoy/grpc/async_client.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <chrono>
+
+#include "envoy/common/optional.h"
+#include "envoy/common/pure.h"
+#include "envoy/grpc/status.h"
+#include "envoy/http/header_map.h"
+
+#include "google/protobuf/descriptor.h"
+
+namespace Envoy {
+namespace Grpc {
+
+/**
+ * An in-flight gRPC stream.
+ */
+template <class RequestType> class AsyncClientStream {
+public:
+  virtual ~AsyncClientStream() {}
+
+  /**
+   * Send request message to the stream.
+   * @param request protobuf serializable message.
+   */
+  virtual void sendMessage(const RequestType& request) PURE;
+
+  /**
+   * Finish the stream. No further operations are permitted on the stream.
+   */
+  virtual void finish() PURE;
+};
+
+/**
+ * Notifies caller of async gRPC stream status.
+ * Note the gRPC stream is full-duplex, even if the local to remote stream has been ended by
+ * AsyncClientStream.finish(), AsyncClientCallbacks can continue to receive events
+ * until the remote to local stream is closed, and vice versa.
+ */
+template <class ResponseType> class AsyncClientCallbacks {
+public:
+  virtual ~AsyncClientCallbacks() {}
+
+  /**
+   * Called when populating the headers to send with initial metadata.
+   * @param metadata initial metadata reference.
+   */
+  virtual void onCreateInitialMetadata(Http::HeaderMap& metadata) PURE;
```

Comment:
This is before the initial metadata is sent, and provides an opportunity to fill in the metadata. This is similar to how `RpcChannel` works today, see https://github.com/lyft/envoy/blob/master/include/envoy/grpc/rpc_channel.h#L30.